presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java

```diff
@@ -556,43 +611,202 @@ private static NestedField nestedField(String name)
+
+                    // If the function accepts lambdas, add all the lambda subfields from each lambda.
+                    Map<Integer, Set<Subfield>> lambdaSubfieldsFromCurrentFunction = ImmutableMap.of();
+                    for (LambdaDescriptor lambdaDescriptor : functionDescriptor.getLambdaDescriptors()) {
+                        Optional<Map<Integer, Set<Subfield>>> lambdaSubfields = collectLambdaSubfields(call, lambdaDescriptor);
+                        if (!lambdaSubfields.isPresent()) {
+                            context.giveUpOnCollectingLambdaSubfields();
+                            call.getArguments().forEach(argument -> argument.accept(this, context));
+                            return null;
+                        }
+                        lambdaSubfieldsFromCurrentFunction = merge(lambdaSubfieldsFromCurrentFunction, lambdaSubfields.get());
+                    }
+
+                    Map<Integer, Set<Subfield>> lambdaSubfields = merge(lambdaSubfieldsFromOuterFunctions, lambdaSubfieldsFromCurrentFunction);
+
+                    lambdaSubfields = addNoSubfieldIfNoAccessedSubfieldsFound(call, lambdaSubfields);
+
+                    // We need to continue visiting the function arguments and collect all lambda subfields in inner function calls as well as non-lambda subfields in all
+                    // function arguments. Once reached the leaf node, we will try to prune the subfields of the input field, subscript, or subfield.
+                    for (int callArgumentIndex = 0; callArgumentIndex < call.getArguments().size(); callArgumentIndex++) {
+                        // Since context is global during the traversal of all the nodes in expression tree, we need to pass lambda subfields only to those  function
+                        // arguments that they relate to.
+                        if (lambdaSubfields.containsKey(callArgumentIndex)) {
+                            context.setLambdaSubfields(lambdaSubfields.get(callArgumentIndex));
+                        }
+                        else {
+                            context.setLambdaSubfields(Context.allSubfieldsOfArrayElementOrMapValue);
+                        }
+                        call.getArguments().get(callArgumentIndex).accept(this, context);
+                    }
+
+                    // When we are done with inner calls (child nodes) we need to restore lambda subfields we received from parent expression to handle such situations like
+                    // in example below
+                    // SELECT * FROM my_table WHERE ANY_MATCH(column1, x -> x.ds > '2023-01-01') AND ALL_MATCH(column2, x -> STRPOS(x.comment,  'Presto') > 0)
+                    // After we are done with ANY_MATCH, we need to restore the lambda subfields to what we received from parent node 'AND' so that it does not collide with
+                    // lambda subfields of ALL_MATCH function.
+                    context.setLambdaSubfields(lambdaSubfieldsOriginal);
```

Comment:
`setLambdaSubfields()` is used in `giveUpOnCollectingLambdaSubfields()`. Since we have ImmutableSet there it is not possible to change the set of subfields when making the field `lambdaSubfields` `final`.

Reply:
Why not? giveUpOnCollectingLambdaSubfields() can also be written as 
```
Context newContext = context. withLambdaSubfields(ALL_SUBFIELDS_OF_ARRAY_ELEMENT_OR_MAP_VALUE)
```